lib/compat.php

```diff
@@ -304,6 +304,263 @@ function gutenberg_warn_classic_about_blocks() {
+	if ( empty( $html ) || empty( $block['attrs'] ) ) {
+		return $html;
+	}
+
+	$defaults = array(
+		'url'   => '',
+		'alt'   => '',
+		'id'    => 0,
+		'align' => '',
+	);
+
+	$block_attributes = wp_parse_args( $block['attrs'], $defaults );
+
+	if ( empty( $block_attributes['url'] ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	if ( ! empty( $block_attributes['id'] ) ) {
+		$attachment_id = (int) $block_attributes['id'];
+		$image_meta    = wp_get_attachment_metadata( $attachment_id );
+	} else {
+		$attachment_id = 0;
+		$image_meta    = null;
+	}
+
+	$image_dimensions = gutenberg_get_image_width_height( $block_attributes, $image_meta );
+	$image_src        = '';
+	$srcset           = '';
+	$sizes            = '';
+
+	if ( empty( $image_dimensions ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	$image_src = $block_attributes['url'];
+
+	$image_attributes = array(
+		'src'    => $image_src,
+		'alt'    => empty( $block_attributes['alt'] ) ? '' : $block_attributes['alt'],
+		'width'  => $image_dimensions[0],
+		'height' => $image_dimensions[1],
+	);
+
+	if ( $image_meta ) {
+		// TODO: pass `$block_attributes` to the filter.
+		$srcset = wp_calculate_image_srcset( $image_dimensions, $image_src, $image_meta, $attachment_id );
+
+		if ( ! empty( $srcset ) ) {
+			// TODO: pass `$block_attributes` to the filter. This will let themes generate better `sizes` attribute.
+			$sizes = wp_calculate_image_sizes( $image_dimensions, $image_src, $image_meta, $attachment_id );
+		}
+
+		if ( $srcset && $sizes ) {
+			$image_attributes['srcset'] = $srcset;
+			$image_attributes['sizes']  = $sizes;
+		}
+	}
+
+	/**
+	 * Filters the image tag attributes when rendering the core image block.
+	 *
+	 * @since 4.5.0
+	 *
+	 * @param array  $image_attributes The (recalculated) image attributes.
+	 *                                 Note: expects valid HTML 5.0 attribute names.
+	 * @param array  $block_attributes The image block attributes.
+	 * @param string $html             The image block HTML coming from the
+	 *                                 editor. The img tag will be replaced.
+	 */
+	$image_attributes = apply_filters( 'render_block_core_image_tag_attributes', $image_attributes, $block_attributes, $html );
+
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names
+		//
+		// See: https://www.w3.org/TR/html52/syntax.html#attribute-names .
+		$is_invalid_attribute = preg_match( '/[\\\\u007F-\\\\u009F "\'>\/="\\\\uFDD0-\\\\uFDEF]/', strtolower( $name ) );
+
+		if ( $is_invalid_attribute ) {
+			continue;
+		}
+
+		if ( 'src' === $name ) {
+			$value = esc_url( $value );
+		} elseif ( ( 'width' === $name || 'height' === $name ) && ! empty( $value ) ) {
+			$value = (int) $value;
+		} else {
+			$value = esc_attr( $value );
+		}
+
+		if ( empty( $value ) ) {
```

Comment:
Boolean attributes are of a whitelisted set, and I don't think it's safe to treat all empty values in the same sense.

For example, this has the impact of writing `img` tags with `alt` omitted as:

![image](https://user-images.githubusercontent.com/1779930/48800204-1dfc5500-ecd7-11e8-925d-5948a7b22655.png)

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

https://github.com/WordPress/gutenberg/blob/5a8c012de43780f7187eecf7ac796e6f57a04c4f/packages/element/src/serialize.js#L92-L136

Reply:
I think it's probably best to avoid this part of the condition and always write the attribute, even if empty. Particularly since it's safe (semantically equivalent) for boolean attributes to "needlessly" have a value assigned.

_Semantically equivalent:_

```html
<input disabled="disabled">
<input disabled="">
<input disabled>
<input disabled="wut">
```